From cece7622e27f80e741bce6139e1ed372163cfed4 Mon Sep 17 00:00:00 2001
From: Asuna Yang <spriteovo@gmail.com>
Date: Thu, 4 Dec 2025 08:54:51 +0100
Subject: [PATCH 032/440] BACKPORT: FROMLIST: rust: export BINDGEN_TARGET from
 a separate Makefile

A subsequent commit will add a new function `bindgen-option` to
`scripts/Kconfig.include`. The bindgen backend requires the `--target`
option for cross compiling, but variable `BINDGEN_TARGET` in
`rust/Makefile` cannot be exported to `scripts/Kconfig.include`.

Therefore, move this variable to a separate new `Makefile.rust` file and
include it from `scripts/Makefile` to make the exported variable
available for use in Kconfig. Place the include in the `need-compiler`
branch to avoid including it in irrelevant make targets.

Since the new file name is `Makefile.rust`, it matches an existing
MAINTAINERS rule `scripts/*rust*`, so no modification to the MAINTAINERS
file is needed.

Signed-off-by: Asuna Yang <SpriteOvO@gmail.com>
Link: https://lore.kernel.org/r/20251204-gcc-rust-v5-v5-1-2d4f20d86c24@gmail.com
Signed-off-by: Han Gao <gaohan@iscas.ac.cn>
Signed-off-by: Mingcong Bai <jeffbai@aosc.io>

[ Mingcong Bai: Resolved a minor post-7.0 merge conflict in
  rust/Makefile ]

Signed-off-by: Mingcong Bai <jeffbai@aosc.io>
---
 Makefile              |  3 ++-
 rust/Makefile         | 10 ----------
 scripts/Makefile.rust | 11 +++++++++++
 3 files changed, 13 insertions(+), 11 deletions(-)
 create mode 100644 scripts/Makefile.rust

diff --git a/Makefile b/Makefile
index d7cca59812d85..e6a994e243e38 100644
--- a/Makefile
+++ b/Makefile
@@ -737,9 +737,10 @@ ifneq ($(findstring clang,$(CC_VERSION_TEXT)),)
 include $(srctree)/scripts/Makefile.clang
 endif
 
+ifdef need-compiler
+include $(srctree)/scripts/Makefile.rust
 # Include this also for config targets because some architectures need
 # cc-cross-prefix to determine CROSS_COMPILE.
-ifdef need-compiler
 include $(srctree)/scripts/Makefile.compiler
 endif
 
diff --git a/rust/Makefile b/rust/Makefile
index bec9726f256ce..45078570e217d 100644
--- a/rust/Makefile
+++ b/rust/Makefile
@@ -398,16 +398,6 @@ bindgen_skip_c_flags := -mno-fp-ret-in-387 -mpreferred-stack-boundary=% \
 	-fdiagnostics-show-context -fdiagnostics-show-context=% \
 	--param=% --param asan-% -fno-isolate-erroneous-paths-dereference
 
-# Derived from `scripts/Makefile.clang`.
-BINDGEN_TARGET_x86	:= x86_64-linux-gnu
-BINDGEN_TARGET_arm64	:= aarch64-linux-gnu
-BINDGEN_TARGET_arm	:= arm-linux-gnueabi
-BINDGEN_TARGET_loongarch	:= loongarch64-linux-gnusf
-# This is only for i386 UM builds, which need the 32-bit target not -m32
-BINDGEN_TARGET_i386	:= i386-linux-gnu
-BINDGEN_TARGET_um	:= $(BINDGEN_TARGET_$(SUBARCH))
-BINDGEN_TARGET		:= $(BINDGEN_TARGET_$(SRCARCH))
-
 # All warnings are inhibited since GCC builds are very experimental,
 # many GCC warnings are not supported by Clang, they may only appear in
 # some configurations, with new GCC versions, etc.
diff --git a/scripts/Makefile.rust b/scripts/Makefile.rust
new file mode 100644
index 0000000000000..fd48b0e0d46fb
--- /dev/null
+++ b/scripts/Makefile.rust
@@ -0,0 +1,11 @@
+# Derived from `scripts/Makefile.clang`.
+BINDGEN_TARGET_x86    := x86_64-linux-gnu
+BINDGEN_TARGET_arm64  := aarch64-linux-gnu
+BINDGEN_TARGET_arm    := arm-linux-gnueabi
+BINDGEN_TARGET_loongarch      := loongarch64-linux-gnusf
+# This is only for i386 UM builds, which need the 32-bit target not -m32
+BINDGEN_TARGET_i386   := i386-linux-gnu
+BINDGEN_TARGET_um     := $(BINDGEN_TARGET_$(SUBARCH))
+BINDGEN_TARGET                := $(BINDGEN_TARGET_$(SRCARCH))
+
+export BINDGEN_TARGET
-- 
2.55.0

